File: frontend/src/app/core/services/reservas.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface DetalleReservaItem {
  variante_id: number;
  cantidad: number;
}

export interface DetalleReserva {
  id: number;
  reserva_id: number;
  variante_id: number;
  cantidad: number;
}

export interface Reserva {
  id: number;
  usuario_id: number;
  sucursal_id: number;
  fecha_reserva: string;
  estado: string;
  observaciones?: string | null;
  detalles: DetalleReserva[];
}

export interface ReservaCrear {
  sucursal_id: number;
  fecha_reserva: string;
  observaciones?: string | null;
  detalles: DetalleReservaItem[];
}

@Injectable({
  providedIn: 'root'
})
export class ReservasService {
  private http = inject(HttpClient);
  private apiUrl = 'https://menstyle-hms1.onrender.com/api/reservas';

  private headers(): HttpHeaders {
    return new HttpHeaders({
      Authorization: `Bearer ${localStorage.getItem('access_token')}`
    });
  }

  listarMisReservas(estado?: string): Observable<Reserva[]> {
    const query = estado ? `?estado=${estado}` : '';
    return this.http.get<Reserva[]>(
      `${this.apiUrl}${query}`,
      { headers: this.headers() }
    );
  }

  obtener(id: number): Observable<Reserva> {
    return this.http.get<Reserva>(
      `${this.apiUrl}/${id}`,
      { headers: this.headers() }
    );
  }

  crear(datos: ReservaCrear): Observable<Reserva> {
    return this.http.post<Reserva>(
      this.apiUrl,
      datos,
      { headers: this.headers() }
    );
  }

  cancelar(id: number): Observable<Reserva> {
    return this.http.put<Reserva>(
      `${this.apiUrl}/${id}/cancelar`,
      {},
      { headers: this.headers() }
    );
  }

  // ----- Administración / encargado de sucursal -----
  listarTodas(estado?: string, sucursalId?: number): Observable<Reserva[]> {
    const params = new URLSearchParams();
    if (estado) params.set('estado', estado);
    if (sucursalId) params.set('sucursal_id', String(sucursalId));
    const query = params.toString() ? `?${params.toString()}` : '';

    return this.http.get<Reserva[]>(
      `${this.apiUrl}/admin/todas${query}`,
      { headers: this.headers() }
    );
  }

  cambiarEstado(id: number, nuevoEstado: string): Observable<Reserva> {
    return this.http.put<Reserva>(
      `${this.apiUrl}/${id}/estado`,
      { nuevo_estado: nuevoEstado },
      { headers: this.headers() }
    );
  }
}

File: mobile/lib/services/reservas_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';
import 'api_client.dart';

class ReservasService {
  static final ReservasService instance = ReservasService._internal();
  factory ReservasService() => instance;
  ReservasService._internal();

  Future<List<Reserva>> listarMisReservas() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/reservas'),
      headers: ApiClient.authHeaders(),
    );
    _verificar(response);
    final data = jsonDecode(response.body) as List;
    return data.map((e) => Reserva.fromJson(e)).toList();
  }

  Future<Reserva> crear({
    required int sucursalId,
    required DateTime fechaReserva,
    String? observaciones,
    required List<Map<String, int>> detalles,
  }) async {
    final response = await http.post(
      Uri.parse('${ApiClient.baseUrl}/reservas'),
      headers: ApiClient.jsonHeaders(),
      body: jsonEncode({
        'sucursal_id': sucursalId,
        'fecha_reserva': fechaReserva.toUtc().toIso8601String(),
        'observaciones': observaciones,
        'detalles': detalles,
      }),
    );
    _verificar(response);
    return Reserva.fromJson(jsonDecode(response.body));
  }

  Future<Reserva> cancelar(int reservaId) async {
    final response = await http.put(
      Uri.parse('${ApiClient.baseUrl}/reservas/$reservaId/cancelar'),
      headers: ApiClient.authHeaders(),
    );
    _verificar(response);
    return Reserva.fromJson(jsonDecode(response.body));
  }

  // ----- Administración -----
  Future<List<Reserva>> listarTodas({String? estado, int? sucursalId}) async {
    final params = <String>[];
    if (estado != null) params.add('estado=$estado');
    if (sucursalId != null) params.add('sucursal_id=$sucursalId');
    final query = params.isNotEmpty ? '?${params.join('&')}' : '';

    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/reservas/admin/todas$query'),
      headers: ApiClient.authHeaders(),
    );
    _verificar(response);
    final data = jsonDecode(response.body) as List;
    return data.map((e) => Reserva.fromJson(e)).toList();
  }

  Future<Reserva> cambiarEstado(int reservaId, String nuevoEstado) async {
    final response = await http.put(
      Uri.parse('${ApiClient.baseUrl}/reservas/$reservaId/estado'),
      headers: ApiClient.jsonHeaders(),
      body: jsonEncode({'nuevo_estado': nuevoEstado}),
    );
    _verificar(response);
    return Reserva.fromJson(jsonDecode(response.body));
  }

  void _verificar(http.Response response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw Exception(_extraerError(response.body));
    }
  }

  String _extraerError(String body) {
    try {
      final data = jsonDecode(body);
      if (data is Map && data['detail'] != null) return data['detail'].toString();
    } catch (_) {}
    return 'No se pudo completar la operación de la reserva';
  }
}
